### PR TITLE
Add undocumented methods and make types more specific in 2/unittest

### PR DESCRIPTION
**Pull request by janverb**
This brings the stubs more in line with the Python 3 stubs and matches the implementation.

`TextTestRunner.run` is mentioned a few times in the [docs](https://docs.python.org/2/library/unittest.html) but it's not listed separately so I think it's technically undocumented.

### CODE REVIEW

File: stdlib/2/unittest.pyi [hunk 1]
```diff
@@ -15,6 +15,10 @@ _FT = TypeVar('_FT')
 _ExceptionType = Union[Type[BaseException], Tuple[Type[BaseException], ...]]
 _Regexp = Union[Text, Pattern[Text]]
 
+_SysExcInfoType = Tuple[Optional[Type[BaseException]],
+                        Optional[BaseException],
+                        Optional[types.TracebackType]]
+
 class Testable(metaclass=ABCMeta):
```
Comment (srittau) [inline stdlib/2/unittest.pyi]: Do we even need this class? It does not exist in the implementation and I don't really know what purpose it serves in the stubs, especially now that you removed its uses.

File: stdlib/2/unittest.pyi [hunk 2]
```diff
@@ -15,6 +15,10 @@ _FT = TypeVar('_FT')
 _ExceptionType = Union[Type[BaseException], Tuple[Type[BaseException], ...]]
 _Regexp = Union[Text, Pattern[Text]]
 
+_SysExcInfoType = Tuple[Optional[Type[BaseException]],
+                        Optional[BaseException],
+                        Optional[types.TracebackType]]
```
Comment (srittau) [inline stdlib/2/unittest.pyi]: Would it make more sense to use `Union[Tuple[None, None, None], Tuple[Type[BaseException], BaseException, types.TracebackType]]`?

File: stdlib/2/unittest.pyi (hunk 1 above)
Comment (janverb) [reply stdlib/2/unittest.pyi]: It effectively means `Union[TestCase, TestSuite]`. I didn't remove all uses, just the ones that I noticed should be `TestCase` instead.

It shouldn't be a class, but changing that would take a lot more work.

File: stdlib/2/unittest.pyi (hunk 2 above)
Comment (janverb) [reply stdlib/2/unittest.pyi]: You're right, that matches `sys.exc_info()`. I copied it from the Python 3 stubs, I'll change it there too.

File: stdlib/2/unittest.pyi [hunk 3]
```diff
@@ -240,12 +244,16 @@ defaultTestLoader: TestLoader
 
 class TextTestResult(TestResult):
     def __init__(self, stream: TextIO, descriptions: bool, verbosity: int) -> None: ...
+    def getDescription(self, test: TestCase) -> str: ...  # undocumented
+    def printErrors(self) -> None: ...  # undocumented
+    def printErrorList(self, flavour: str, errors: List[Tuple[TestCase, str]]) -> None: ...  # undocumented
```
[the remark this answers is not in the document]
Comment (janverb) [reply stdlib/2/unittest.pyi]: The implementation calls it with `TestResult.errors` (`List`) as an argument. Making it `List` allows implementations in subclasses to assume the argument is a list, which I think is more useful given that it's routinely called by `printErrors` and there's not much reason to call it from the outside.

**Comment by rchen152** [comment]
I'm trying to import a new version of typeshed into Google, and this change is causing new type errors in Python 2. What's happening is that we have code calling `TestResult.addUnexpectedSuccess(<instance of FunctionTestCase>)`, which now fails to type check because `addUnexpectedSuccess` and friends were changed to accept `TestCase` instead of `Testable`.

In the unittest implementation, `FunctionTestCase` is a subclass of `TestCase`: https://github.com/python/cpython/blob/249706c1fbce04125d81bd9993e6c010ae30f8e4/Lib/unittest/case.py#L1019.

In unittest.pyi, `FunctionTestCase` and `TestCase` both subclass `Testable`. (I'm guessing this choice was made because the documentation says that `FunctionTestCase` implements only a portion of the `TestCase` interface: https://docs.python.org/2/library/unittest.html#unittest.FunctionTestCase.)

It seems to me that the type definition of `TestResult.addUnexpectedSuccess` has been changed to follow the implementation, but the type definitions intentionally differ slightly from the implementation, so we should either (1) change the function signature to accept `Testable` again or (2) make `FunctionTestCase` inherit from `TestCase`. Thoughts?

**Comment by JelleZijlstra** [comment]
(2) sounds right to me; if it inherits from TestCase at runtime the stub should also inherit from TestCase.

**Comment by rchen152** [comment]
Sounds good, I'll make a PR.

**Pull request merged**